python/lsst/ip/diffim/detectAndMeasure.py

```diff
@@ -453,33 +425,61 @@ def _deblend(self, difference, positiveFootprints, negativeFootprints):
-            footprints = afwDetection.FootprintSet(difference.getBBox())
-            footprints.setFootprints([src.getFootprint() for src in sources])
-            return footprints
-
-        def deblend(footprints):
-            """Deblend a positive or negative footprint set,
-            and return the deblended children.
-            """
-            sources = afwTable.SourceCatalog(self.schema)
-            footprints.makeSources(sources)
-            self.deblend.run(exposure=difference, sources=sources)
-            self.setPrimaryFlags.run(sources)
-            children = sources["detect_isDeblendedSource"] == 1
-            sources = sources[children].copy(deep=True)
-            # Clear parents, so that measurement plugins behave correctly.
-            sources['parent'] = 0
-            return sources.copy(deep=True)
-
-        positives = deblend(positiveFootprints)
-        negatives = deblend(negativeFootprints)
-
-        sources = afwTable.SourceCatalog(self.schema)
-        sources.reserve(len(positives) + len(negatives))
-        sources.extend(positives, deep=True)
-        sources.extend(negatives, deep=True)
-        return sources, makeFootprints(positives), makeFootprints(negatives)
+        # Merge the positive and negative footprints, but don't grow them.
+        merged_footprints = positiveFootprints
+        merged_footprints.merge(negativeFootprints, 0, 0, False)
+
+        # Create a source catalog from the footprints.
+        table = afwTable.SourceTable.make(self.schema, idFactory)
+        sources = afwTable.SourceCatalog(table)
+        merged_footprints.makeSources(sources)
+
+        # Sky sources must be added before deblending, otherwise the
+        # sources with parent == 0 will be out of order and SFM cannot run.
+        if self.config.doSkySources:
+            self.addSkySources(sources, difference.mask, difference.info.id)
+
+        # Find the footprints with multiple positive peaks no negative peaks.
+        footprints = [src.getFootprint() for src in sources]
+        nPeaks = np.array([len(fp.peaks) for fp in footprints])
+        blend_ids = []
+        skipped_ids = []
+        for src in sources[nPeaks > 1]:
+            peaks = src.getFootprint().peaks
+            peak_flux = np.array([peak.getPeakValue() for peak in peaks])
+            if np.all(peak_flux > 0):
+                blend_ids.append(src.getId())
+            else:
+                skipped_ids.append(src.getId())
+
+        # Deblend the footprints that can be deblended with SourceDeblendTask.
+        temp_cat = afwTable.SourceCatalog(sources.table)
+        for sid in blend_ids:
+            temp_cat.append(sources.find(sid))
+        first_child_index = len(temp_cat)
+        self.deblend.run(exposure=difference, sources=temp_cat)
+
+        # Update the deblended parents and add their children
+        # to the sources catalog.
+        sources.extend(temp_cat[first_child_index:], deep=False)
```

Comment:
I think you want to do a deep copy here to ensure the catalog is contiguous, but maybe that's not necessary for the measurement tasks that are run in after this? And then the catalog is written to disk. I guess we can see if we get failures when running on larger data?

Reply:
No, we actually don't want a deep copy (see Jim's comments in slack). The temporary catalog actually created a view to the original records, that way the source IDs stay in order and the original records are updated in place.

Reply:
I checked and verified that using this procedure the sources are always contiguous.